internal/transformations/static_analyzer.go

```diff
@@ -164,6 +164,9 @@ func analyzeVisit(a ast.Node, inObject bool, vars ast.IdentifierSet) error {
 	return s.err
 }
 
-func analyze(node ast.Node) error {
+// Analyze checks variable references (these could be checked statically in Jsonnet).
+// It enriches ast with additional information about free variables in every node,
+// so it is necessary to always run it before executing AST.
+func Analyze(node ast.Node) error {
```

Comment:
That this function is exported from this package suggests that neither "transformations" nor "transform" capture all of what's going on in here. The documentation suggests that this function mutates the supplied AST in place.

Given that one is not supposed to execute an AST until it's been analyzed, we could enforce that by having this function return a type that tolerates such execution, and disallow doing that directly against a raw AST. Unfortunately, that would break backward compatibility.

Instead of offering `(*VM).Evaluate(ast.Node)` and kin, we could call this package "program" introduce a type like `program.Top` that wraps a prepared `ast.Node`.
```go
func Prepare(node ast.Node) (Top, error) {
```
We'd then rephrase `(*VM).Evaluate(ast.Node)` and kin as `(*VM).Evaluate(program.Top)`. In other words, you couldn't feed an `ast.Node` into the evaluation procedures, because it hadn't been properly prepared.

Reply:
Hmmm... I like the name "program". Even without introducing a new type it makes sense (it prepares the program to be run). It's also short and nice.

For now I'm leaving out the new type but I agree with the argument for it. It just seems out of scope here and it I'm not sure if backwards compatibility issues in this case will be acceptable.